Task 1

Saxophilyman opened the pull request.

Saxophilyman committed:
Create Java Project(empty)

Saxophilyman committed:
Part_I "Hello message"

Saxophilyman committed:
Part_II "First Steps(sketch)"

Saxophilyman committed:
Part_II "First Steps(sketch)"
Hope that this shit also added

Saxophilyman committed:
What is logic? (also add exit and repeat Begin in console)

Saxophilyman committed:
What is logic? (also add exit and repeat Begin in console)

Saxophilyman committed:
Part_III "Fucking Fuck"
Different lost ways

Saxophilyman committed:
Part_III "Details"
 - clear excess;
 - add readConsole;
 - add addCoord;
 - working on checkEndGame

Saxophilyman committed:
Part_III "Details 1.1" (thoughts about two player)

Saxophilyman committed:
Part_IV "Trying OOP"
 - rename class Begining to Game;
 - delete other classes;
 - clear Main(go in Game);
 - create class Player and his methods(+ 2 object);
 - some logic of game with class Player;

 - working on checkEndGame

Saxophilyman committed:
Part_V "EndProblem"
 - addedCord
 - notEqualsaddedCord

Saxophilyman committed:
Part_VI "Trying to be ready"
 - delete Player.java
 - true for checkEndGame
 - while & logic for Playing
 - Begin() in Main
 - circle gameGoOn(with Begin)
 - refreshGamePole

 - delete excess

Saxophilyman committed:
"Trying git_ignore"

Saxophilyman committed:
"Trying git_ignore 2.0"

epereskokov commented:
Комментарий касательно всего кода:
Есть документ, называется Java Code Conventions (https://www.oracle.com/technetwork/java/codeconventions-150003.pdf) - это документ-соглашение между разработчиками джава, как оформлять код. Я тебе буду правки по код стайлу писать согласно этому соглашению, т.е. если я говорю, что оформлять надо так и именовать так - это не моя прихоть, а де-факто норма оформления.
Итак:
1) В IDEA есть автоформатирование кода (ctrl+alt+L), форматирует текущий файл. Зайди в оба своих файла с кодом и отформатируй.
2) именование:
- классов - Camel Case с большой буквы - например MyFirstSuperClass
- методов - Camel Case с маленькой буквы, например mySuperImprotantMethod(). Метод должен содержать глагол, описывающий суть действия.
- переменных - Camel Case с маленькой буквы 
- если переменная - константа (static final), то используется Upper Case через нижние подчеркивания, например MY_FIRST_CONSTANT
- В именовании классов, переменных и методов настоятельно рекомендуется использовать только буквы, цифр, нижних подчеркиваний и т.п. быть не должно (кроме нижнего подчеркивания в названии констант)
3) Инкапсуляция (сокрытие реализации) - один из принципов ООП (объектно-ориентированное программирование), в джаве он реализуется модификаторами доступа для классов, методов и переменных (public, protected, package-private и private). Суть принципа - в предоставлении минимально необходимого доступа к внутренней реализации твоего кода внешним пользователям. Иными словами - все что может быть private - обязательно должно быть private.  Пройдись по коду и расставь минимально возможные модификаторы доступа.
Переменные класса должны быть всегда private, доступ к ним должен осуществляться через get и set-методы, пример:

_неправильно:_
```
public class Example {
  public int firstVariable;
}
```

_Правильно:_
```
public class Example {
  private int firstVariable;

  public int getFirstVariable() {
    return firstVariable;
  }

  public void setFirstVariable(int value) {
    this.firstVariable = value;
  }
}
```


Я не буду писать замечание в каждом месте по код-стайлу, пройдись по пунктам выше и поправь свой код. Если ты хочешь поменять имя переменной сразу во всех местах ее использования - поставь курсор на переменную и нажми shit+F6, потом введи новое имя и нажми Enter, переменная изменится сразу во всех местах.

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,166 @@
+            String c2 = GamePole[3][2];
+            String c3 = GamePole[3][3];
+
+            if ((a1.equals(this.mark) && a2.equals(this.mark) && a3.equals(this.mark))||
```
если бы поле было 4х4, то тут было бы куда больше вариантов. А 20х20? Вообще мрак. Тебе нужно извлечь из строки из 2 символов отдельно первый и отдельно второй символ, и работать с ними отдельно. Для этого есть методы у строки либо toCharArray(), либо charAt(), как их использовать - разберешься. (Чтобы посмотреть реализацию метода - зажми ctrl и щелкни левой кнопкой мыши на методе - ты перейдешь в его реализацию).

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,166 @@
+                End = false;
+            }
+
+            else if (addedCord.size()>8){
```
можно обойтись просто int-счетчиком ходов.

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,166 @@
+            assert XO != null;
+            if (XO.matches("a1|a2|a3|b1|b2|b3|c1|c2|c3")){
+            System.out.println("ok");
+                if (notEqualsaddedCord(XO)){
```
чем хранить доп. список, можно просто перед вставкой символа в нужную ячейку двумерного массива проверить, что она пустая, если нет - то значит ход недопустимый.

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,166 @@
+        }
+
+        public boolean notEqualsaddedCord (String x) {
+            boolean B = true;
```
избегай именование переменных в одну букву - всегда давай осмысленные названия - так твой код не будет нуждаться в комментариях и документации (концепция самодокументирующегося кода)

Saxophilyman replied:
ready

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,166 @@
+
+public class Game {
+    public static BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
+    public final static String[][] Pole =  {{"/","1","2","3"},{"a"," "," "," "},{"b"," "," "," "},{"c"," "," "," "}};
```
Для поля есть английское слово Field, не нужно транслитераций с русского.
Массив данных для игры не должен содержать ничего лишнего - все эти визуальные декорации нужно добавлять на этапе вывода в консоль, иначе рано или поздно у тебя перепутаются реальные данные и "декоративные".

Saxophilyman replied:
ready

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,166 @@
+            this.nameMark = nameMark;
+        }
+
+        public void message(){
```
название метода должно говорить о том, что он делает. Здесь например подходящим названием будет printNextPlayerTurnMessage()

Saxophilyman replied:
ready

epereskokov commented:
Двумерный динамический массив - это List<List<String>> yourDwoDimensionArray

Saxophilyman committed:
Part_VII "Trying to be ready 2.0"
 - check pull requests's comment #1

Saxophilyman committed:
Part_VII "Trying to be ready 2.1"
 - check pull requests's comment about decoration

Saxophilyman committed:
Part_VII "Trying to be ready 2.1.1"
 - check add Coord
 - problem with check EndGame in short way

Saxophilyman committed:
Part_VII "Trying to be ready 3.0"
- that's all

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,194 @@
+    Player playerFirst = new Player("Player_1", "X", "крестика");
+    Player playerSecond = new Player("Player_2", "O", "нолика");
+
+    public static class Player {
```
избегай вложенных классов - они кроме усложнения чтения кода абсолютно ничего не дают. На джавараше они использовались исключительно для того, чтобы уместить одно задание в один файл. Тут нужно вынести Player в отдельный класс.

Saxophilyman replied:
ready

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,194 @@
+    private static String[][] workField = {{" ", " ", " "}, {" ", " ", " "}, {" ", " ", " "}};
+    private static String[][] decorField = {{"/", "1", "2", "3"}, {"a"}, {"b"}, {"c"}};
+    private static ArrayList<String> addedCord = new ArrayList<>();
+    Player playerFirst = new Player("Player_1", "X", "крестика");
```
Принимай игроков как параметры в конструкторе. Самих игроков создавай перед началом игры.

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,194 @@
+public class Game {
+    private static BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
+    private final static String[][] GAME_FIELD = {{" ", " ", " "}, {" ", " ", " "}, {" ", " ", " "}};
+    private static String[][] workField = {{" ", " ", " "}, {" ", " ", " "}, {" ", " ", " "}};
```
статичными все эти поля делать нет смысла.

Saxophilyman replied:
ready

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,194 @@
+
+        }
+
+        private boolean checkIfNotEqualsAddedCord(String x) {
```
все методы кроме printMessageMoveOfPlayer() явно должны принадлежать классу Game, ведь объект игры управляет ходом самой игры. Класс Player у тебя является небольшим классом, который описывает игроков (их имя и символ).

Saxophilyman replied:
ready

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,194 @@
+           char [] firstAndSecondCoordReadConsole = a.toCharArray();
+            int firstCoordWorkFieldVertical = firstAndSecondCoordReadConsole[0];
+            int secondCoordWorkFieldHorizontal = firstAndSecondCoordReadConsole[1];
+            if (firstCoordWorkFieldVertical == 97){
```
если ты хочешь сравнить символы - то ты можешь не писать из коды, а использовать одинарные кавычки для описания символов (char), например:
 firstCoordWorkFieldVertical == 'a'

Saxophilyman replied:
ready

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,194 @@
+        }
+    }
+
+    public void playingGame() {
```
startGame()

Saxophilyman replied:
ready

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,194 @@
+    }
+
+
+    public static void printAllGamePole() {
```
printGameField()

Saxophilyman replied:
ready

epereskokov commented on src/Main.java:
```diff
@@ -0,0 +1,64 @@
+        beginingGame();
+    }
+
+    private static void beginingGame() {
```
startNewGame / beginGame

Saxophilyman replied:
ready

epereskokov commented on src/Main.java:
```diff
@@ -0,0 +1,64 @@
+        beginingGame();
+    }
+
+    private static void playingFullGame() {
```
playTicTacToe / playCrossesZeroes

Saxophilyman replied:
ready

epereskokov commented on src/Game.java:
```diff
@@ -0,0 +1,194 @@
+
+public class Game {
+    private static BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
+    private final static String[][] GAME_FIELD = {{" ", " ", " "}, {" ", " ", " "}, {" ", " ", " "}};
```
тебе не обязательно явно инициализировать пустыми строками массив, ты можешь просто создать пустой через new String[3][3], тогда внутри все ячейки по дефолту будут null.

Saxophilyman committed:
Part_VIII "Trying to be ready 3.1"
 - with class Player

Saxophilyman committed:
Part_IX "3.2"
 - add constructor Game(Player, Player)
 - check mark

Saxophilyman committed:
Part_IX "3.3"
 - correct array (use new String[][])
 - correct printGameField(check NULL)
 - correct checkEndGame(reverse equals)
